### DISCUSSION TITLE
version column is null for all rows

### DISCUSSION THREAD

**Discussion by oliverw** [Azure/azure-mobile-apps#607] [Q&A]
I'm using `Microsoft.AspNetCore.Datasync.EFCore` with Postgres 15 and I've noticed that the `version` column is null for all of my entities in the database. Am I missing something? The docs indicate that the this field is maintained by the library. 

> These fields are maintained by the service and shouldn't be set by the client application.

**Comment by adrianhall** [comment]
The Version property is decorated with the [Timestamp] in the EFCore driver, which should maintain the version within the database.  I haven't tried PostgreSQL, so I cannot confirm this behavior.

If that's the case, then you may need to do some more setup in your DbContext.  In your case, a little bit of research suggests that Postgres uses a uint as a timestamp, so you may need to provide some code to ensure that the version is updated on writes appropriately.

**Reply by oliverw** [reply]
> If that's the case, then you may need to do some more setup in your DbContext. In your case, a little bit of research suggests that Postgres uses a uint as a timestamp, so you may need to provide some code to ensure that the version is updated on writes appropriately.

Hmm. It is currently mapped as `bytea` column because `EntityTableData` defines the `Version` property as `byte[]`. Probably because that's how SQL Server defines timestamps.

**Reply by oliverw** [reply]
Judging by the [npgsql docs](https://www.npgsql.org/efcore/modeling/concurrency.html?tabs=data-annotations), it should work out of the box if the `Version` property was an `uint` rather than a `byte[]`.  So in absence of a class `EntityTableData<TVersion> : ITableData` I need to roll my own `EntityTableData` equivalent I guess.

**Reply by adrianhall** [reply]
Using a uint is wrong for other reasons (it's not sufficiently wide to be used as a version moniker).

My recommendation is to introduce something into the DbContext such that the field is set on write.  Probably institute a trigger on a create/update that sets the field to a set of guid bytes.

See this: https://stackoverflow.com/questions/47281414/entity-framework-core-rowversion-column-not-updating-using-postgresql which has some suitable code.

Something like this in the OnModelCreating:

```csharp
    var converter = new ValueConverter<byte[], long>(
        v => BitConverter.ToInt64(v, 0),
        v => BitConverter.GetBytes(v));

    builder.Entity<Author>()
            .Property(_ => _.RowVersion)
            .HasColumnName("xmin")
            .HasColumnType("xid")
            .ValueGeneratedOnAddOrUpdate()
            .HasConversion(converter);
```

**Reply by adrianhall** [reply]
If this works, BTW, let me know as I would love to update the docs.

**Reply by oliverw** [reply]
> Using a uint is wrong for other reasons (it's not sufficiently wide to be used as a version moniker).
> 
> My recommendation is to introduce something into the DbContext such that the field is set on write. Probably institute a trigger on a create/update that sets the field to a set of guid bytes.
> 
> See this: https://stackoverflow.com/questions/47281414/entity-framework-core-rowversion-column-not-updating-using-postgresql which has some suitable code.
> 
> Something like this in the OnModelCreating:
> 
> ```cs
>     var converter = new ValueConverter<byte[], long>(
>         v => BitConverter.ToInt64(v, 0),
>         v => BitConverter.GetBytes(v));
> 
>     builder.Entity<Author>()
>             .Property(_ => _.RowVersion)
>             .HasColumnName("xmin")
>             .HasColumnType("xid")
>             .ValueGeneratedOnAddOrUpdate()
>             .HasConversion(converter);
> ```

I independently came up with pretty much the same solution. Testing..

**Reply by oliverw** [reply]
This results in a rather curious migration attempt to rename the column to the postgres system managed column `xmin`.

```c#
public partial class PgRowVersion : Migration
{
    /// <inheritdoc />
    protected override void Up(MigrationBuilder migrationBuilder)
    {
        migrationBuilder.RenameColumn(
            name: "version",
            table: "tasks",
            newName: "xmin");

        migrationBuilder.AlterColumn<uint>(
            name: "xmin",
            table: "tasks",
            type: "xid",
            rowVersion: true,
            nullable: true,
            oldClrType: typeof(byte[]),
            oldType: "bytea",
            oldRowVersion: true,
            oldNullable: true);
    }

    /// <inheritdoc />
    protected override void Down(MigrationBuilder migrationBuilder)
    {
        migrationBuilder.RenameColumn(
            name: "xmin",
            table: "tasks",
            newName: "version");

        migrationBuilder.AlterColumn<byte[]>(
            name: "version",
            table: "tasks",
            type: "bytea",
            rowVersion: true,
            nullable: true,
            oldClrType: typeof(uint),
            oldType: "xid",
            oldRowVersion: true,
            oldNullable: true);
    }
}

```

**Reply by oliverw** [reply]
The approach works, but you have to manually edit the migration and turn both Up() and Down() into empty methods which is kinda ugly. 

`EntityTableData<TVersion> : ITableData` would be much nicer in my opinion. 

> it's not sufficiently wide to be used as a version moniker

Care to elaborate? 😄

**Reply by oliverw** [reply]
Tried to roll my own base entity implementation below but got stopped [dead in my tracks](https://github.com/Azure/azure-mobile-apps/blob/ba92373099a94e1a94d3318bed02dd3b66e31945/sdk/dotnet/src/Microsoft.AspNetCore.Datasync.EFCore/EntityTableRepository.cs#L38) 😁 

```c#
public class PgEntityTableData : 
    ITableData
{
    [Key]
    public string Id { get; set; }

    /// <summary>
    /// The date/time that the entity was updated.
    /// </summary>
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.UtcNow;

    /// <summary>
    /// The row version for the entity.
    /// </summary>
    [JsonIgnore]
    [NotMapped]
    public byte[] Version
    {
        get => BitConverter.GetBytes(XMin);
        set => throw new NotSupportedException("Postgres xmin-backed version is read-only");
    }

    /// <summary>
    /// True if the entity is marked as deleted.
    /// </summary>
    public bool Deleted { get; set; }

    /// <summary>
    /// Implements the <see cref="IEquatable{T}"/> interface to determine
    /// if the system properties match the system properties of the other
    /// entity.
    /// </summary>
    /// <param name="other">The other entity</param>
    /// <returns>true if the entity matches and the system properties are set.</returns>
    public bool Equals(ITableData other)
        => other != null
           && Id == other.Id
           && UpdatedAt == other.UpdatedAt
           && Deleted == other.Deleted
           && Version.SequenceEqual(other.Version);

    /// <summary>
    /// The actual version
    /// </summary>
    [JsonProperty("version")]
    [Timestamp]
    public UInt64 XMin { get; }
}

```

**Reply by adrianhall** [reply]
Not tested (I don't have Postgres), but:

```csharp
public PgTableData : EntityTableData
{
    [NotMapped]
    public override byte[] Version
    {
        get => BitConverter.GetBytes(XMin);
        set => Debug.WriteLine($"Warning: Ignoring setting of the Version field for ID {this.Id}");
    }

    [JsonIgnore, Timestamp]
    public UInt64 XMin { get; }
}
```

I'd probably set up the set as well, just like the ValueConverter - but this should be enough.

**Reply by oliverw** [reply]
Too bad not possible. `Version` is not virtual.

**Reply by adrianhall** [reply]
Of course - if you pass in an object (across the wire) that has Version, it is set in the model, but it is generally handled by EF.  It's worthwhile doing the set to do the bitconverter as well.

**Reply by adrianhall** [reply]
If you want to work with me on this... compile the entire repo again, but make Version virtual and adjust it.

if that solves the problem (and it's the only problem), then we can easily do a PR to make that change.

**Reply by oliverw** [reply]
> If you want to work with me on this... compile the entire repo again, but make Version virtual and adjust it.
> 
> if that solves the problem (and it's the only problem), then we can easily do a PR to make that change.

Sure.

**Reply by oliverw** [reply]
This appears to work, but I'd appreciate if you could provide a couple manual test scenarios to perform to be sure. So far I have only tested insert, updates and deletes.

```c#
public class PgEntityTableData : EntityTableData
{
    /// <summary>
    /// The row version for the entity.
    /// </summary>
    [NotMapped]
    public override byte[] Version
    {
        get => BitConverter.GetBytes(RowVersion);
        set => BitConverter.ToUInt32(value);
    }

    /// <summary>
    /// The actual version
    /// </summary>
    [JsonIgnore]
    [Timestamp]
    [DatabaseGenerated(DatabaseGeneratedOption.Computed)]
    [Column("xmin", TypeName = "xid")]
    public uint RowVersion { get; set; }
}

```

**Reply by oliverw** [reply]
Any update on this?

**Reply by adrianhall** [reply]
As I mentioned, I do not have a PostgreSQL setup - however, if insert, update, and deletes work (as in the xmin/xid columns get appropriately updated and inserts/updates/soft-deletes all update/change the version value), then it's working.

**Reply by oliverw** [reply]
> As I mentioned, I do not have a PostgreSQL setup - however, if insert, update, and deletes work (as in the xmin/xid columns get appropriately updated and inserts/updates/soft-deletes all update/change the version value), then it's working.

Yeah I know 😄 Just wanted to clarify how we proceed from here on.

**Reply by oliverw** [reply]
By the way `EntityTableData.Version` appears to be not virtual in 5.1.1.

**Reply by adrianhall** [reply]
Ack - File a bug for that, and I'll make it virtual in the next release.

**Reply by oliverw** [reply]
> Ack - File a bug for that, and I'll make it virtual in the next release.

Done.
